DNSCache: TTL cleanup implementation.

eloycoto opened the pull request:
The reason for this commit is to avoid large list of identities on fqdn policy
and implement the cleanup process to clean expires IPS and avoid large times on
policyCalculation.

The reason to use a cleanup map in dnsCache is to avoid to loop all the DNS
cache entries, so each second we have a map and a controller will clean the
entries that has some ips expired at that point of time.

The reason to use the controller, is to avoid to have a goroutine with a
timer.Tick each second, and have all in the same goroutine. To avoid leaks,
instead the passing time.Now() use the cache.lastCleanup to get all the entries
from the last run, policy regeneration can take more than 40 seconds in some
huge policies and with this method leak seconds entries int he cleanup map will
not happen.

Tried also with a callback function, but that can block the cache in case of the
long policyCalculation times and DNS proxy will not resolve/response correctly.
Also tried with the channel, but need to have a buffer and the Cilium way is
more related to use a controller.

This PR will not work if https://github.com/cilium/cilium/pull/6692 is not merged. 

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6740)
<!-- Reviewable:end -->

eloycoto committed:
DNSCache: TTL cleanup implementation.

The reason for this commit is to avoid large list of identities on fqdn policy
and implement the cleanup process to clean expires IPS and avoid large times on
policyCalculation.

The reason to use a cleanup map in dnsCache is to avoid to loop all the DNS
cache entries, so each second we have a map and a controller will clean the
entries that has some ips expired at that point of time.

The reason to use the controller, is to avoid to have a goroutine with a
timer.Tick each second, and have all in the same goroutine. To avoid leaks,
instead the passing time.Now() use the cache.lastCleanup to get all the entries
from the last run, policy regeneration can take more than 40 seconds in some
huge policies and with this method leak seconds entries int he cleanup map will
not happen.

Tried also with a callback function, but that can block the cache in case of the
long policyCalculation times and DNS proxy will not resolve/response correctly.
Also tried with the channel, but need to have a buffer and the Cilium way is
more related to use a controller.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

aanm commented on daemon/fqdn.go:
```diff
@@ -100,6 +101,17 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
 
+	// Controller to cleanup TTL expired entries from the DNS policies.
+	controller.NewManager().UpdateController("dns-ttl-cleanup-job", controller.ControllerParams{
+		RunInterval: time.Second,
```
I'm probably the wrong person to review this PR but do we need to run this controller every second? Can't this be event based on the TTL of the entry?

eloycoto replied:
no, each second the cleanup need to happens to clean the table.

aanm replied:
But the entries don't have a TTL for every second

eloycoto replied:
Multiple dns queries has multiple DNS TTL, so each second we need to clean the TTL that has expired after X seconds.

raybejjani replied:
The challenge is knowing when to wake up to expire a specific IP and regenerate policy. We know the expiration time of each entry and setting a timer for the next one to expire is straight-forward Figuring out the expire time for the one after the next is annoying. We would need something like a priority queue (with O(logn) best case update). In theory, most updates will add a new time to the end of this queue (i.e. a new latest time). Still, we can have situations where new entries will expire somewhere in the middle.
Binning the entries into seconds seems like the lowest overhead, since we have as many map entries as we have IPs (or less). The lookup is also constant time for a specific second.

tgraf replied:
I doubt that we need 1 second accuracy for the TTL. Isn't it enough to expire every 15 seconds?

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,42 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+// CleanupExpiredEntries cleans all the expired entries from the lastTime that
+// runs to the give time. It will lock the struct until retrieves all the data.
+// It returns the list of names that need to be deleted from the policies.
+func (c *DNSCache) CleanupExpiredEntries(expires time.Time) []string {
```
Let's return lastCleanup as well, this way it will be clear what the time range was for logging etc.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,42 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+	timediff := int(expires.Sub(c.lastCleanup).Seconds())
+	expiredEntries := []string{}
+
+	for i := 0; i < int(timediff); i++ {
```
This is fine but we can also use `for c.lastCleanup.Before(expires) {` for the loop condition.

eloycoto replied:
I prefer to use the timediff tbh, is a bit clear what is trying to achieve.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,42 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+		expiredEntries = append(expiredEntries, entries...)
+		delete(c.cleanup, key)
+	}
+	return expiredEntries
```
Since we now do the actual removal of cacheEntry objects here, we need to do a :
```
	for _, name := range expiredEntries {
		if entries, exists := c.forward[name]; exists {
			entries.removeExpired(expiredTime)
		}
	}
```

We also might want to sort and unique the names. If we do both, the loop above already does the uniqueing because of the exists check (if we add to a list only when exists == true).

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -406,3 +406,13 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	// Sleep to make sure that it's deleted
+	time.Sleep(5 * time.Second)
+	c.Assert(len(cache.CleanupExpiredEntries(time.Now())), Equals, 1)
+	c.Assert(len(cache.cleanup), Equals, 0)
```
Also maybe check that the normal lookup returns no data to complete the loop.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -108,15 +108,23 @@ type DNSCache struct {
+	lastCleanup time.Time
+
+	// cleanup is the map of TTL expiration time and the entries that are going
+	// to expire at that second.
```
mention that the expiration time is in seconds, and that it is expected that we will delete entries in `CleanupExpiredEntries`

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,42 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
-	// When lookupTime is much earlier than time.Now(), we may not expire all
-	// entries that should be expired, leaving more work for .Lookup.
-	c.removeExpired(entries, time.Now())
+	c.addNameToCleanup(entry)
```
This line needs to be in updateWithEntryIPs. This is because we're tracking by IP, but store the whole DNS lookup. The entry may not actually go into the cache if, for a particular IP, there is another lookup that expires later.
In the cases were, for an IP, we are replacing an existing entry, we might need to also replace the cleanup entry. This is because that IP won't actually expire at the original time anymore. It is a little harmless, though.

raybejjani commented on daemon/fqdn.go:
```diff
@@ -100,6 +101,17 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
 	}
 
+	// Controller to cleanup TTL expired entries from the DNS policies.
+	controller.NewManager().UpdateController("dns-ttl-cleanup-job", controller.ControllerParams{
```
You were right: putting this here makes things a little complicated. We now need a controller for the Endpoint.DNSHistory caches(unless some of the changes below around replacing cleanup entries go in, then it might end up with a steady memory usage).

raybejjani reviewed (changes requested):
One thing to change, otherwise LGTM!

aanm commented on pkg/fqdn/cache_test.go:
```diff
@@ -406,3 +406,13 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	cache.Update(now, "test.com", []net.IP{net.ParseIP("1.2.3.4")}, 3)
+	c.Assert(len(cache.cleanup), Equals, 1)
+	// Sleep to make sure that it's deleted
+	time.Sleep(5 * time.Second)
```
to avoid sleeping, in the next slide you should be able to do `cache.CleanupExpiredEntries(time.Now().Add(5*time.Second))..`

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,56 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+			c.removeExpired(entries, c.lastCleanup)
+		}
+	}
+	return uniqExpires, c.lastCleanup
```
Oh, I meant the original lastCleanup at the start of this function, not the final one (which is going to be roughly `expires`).

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,56 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
+	expiration := int64(entry.TTL) + time.Now().Unix()
+	expiredEntries, exists := c.cleanup[expiration]
+	if !exists {
```
I just realised that the append means we can avoid this if, if you prefer

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,56 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+// AddNameToCleanup adds the IP with the given TTL to the the cleanup map to
+// delete the entry from the policy when it expires.
+// Need to be called with the lock
+func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
```
We should avoid adding this to cleanup if `entry.ExpirationTime` is before `c.lastCleanup`

eloycoto replied:
How can that be possible? the TTL is time.Now() + TTL no?

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,9 +162,56 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+// delete the entry from the policy when it expires.
+// Need to be called with the lock
+func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
+	expiration := int64(entry.TTL) + time.Now().Unix()
```
I just realised that this isn't right. We already set the ExpirationTime in the entry and use that everywhere else. We should probably use it here too, so it is consistent. It will also apply the TTL correctly relative to the lookup time of the data, instead of `time.Now()` (which will be within the same second most of the time so it would mostly work this way).

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -268,6 +323,7 @@ func (c *DNSCache) updateWithEntryIPs(entries ipEntries, entry *cacheEntry) {
 			c.upsertReverse(ipStr, entry)
 		}
 	}
+	c.addNameToCleanup(entry)
```
This needs to be called when we call `c.upsertReverse` (just above) otherwise we introduce expirations that aren't actually happening in the cache.

tgraf commented on daemon/fqdn.go:
```diff
@@ -100,6 +101,17 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
+			d.dnsRuleGen.ForceGenerateDNS(namesToClean)
+			return nil
+		},
+		StopFunc: func() error { return nil },
```
This can be omitted

eloycoto committed:
FQDN: Added garbage collector functions.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

raybejjani commented on daemon/fqdn.go:
```diff
@@ -106,17 +107,42 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
+			for i, ep := range endpoints {
+				cfg.Cache.UpdateFromCache(ep.DNSHistory, namesToClean, i == 0)
+			}
+			log.Infof("FQDN garbage collector work deleted %d name entries", len(namesToClean))
```
this is probably a debug... I'm not sure how relevant it is without the list of names deleted

eloycoto replied:
Just a info message to be able to track this down in the future if any issue.

raybejjani replied:
shrug. People keep complaining to me about log lines.
Anyway, can you make it a scopedLog with controller name or whatever we usually put into the controllers? that will make it easier to filter for it.

raybejjani commented on daemon/fqdn.go:
```diff
@@ -106,17 +107,42 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState) (err err
+			}
+			// A second loop is needed to update the global cache from the
+			// endpoints cache
+			for i, ep := range endpoints {
```
Also mention: The first update from an endpoint will clear the data in cfg.Cache for each of namesToClean. Looping this way is generally safe despite not locking; If a new lookup happens during these updates the new DNS data will be reinserted from the endpoint.DNSHistory cache that made the request.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -108,15 +111,39 @@ type DNSCache struct {
+	cleanup map[int64][]string
+
+	// overLimit is a map with the host names that has reached the ips limits
+	// defined by the DNSCache
```
defined by limit, no?

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +181,124 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 		c.forward[entry.Name] = entries
 	}
 	c.updateWithEntryIPs(entries, entry)
+	if len(entries) > FQDNMaxIPperHost {
```
this now needs to use c.limit

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -108,15 +111,39 @@ type DNSCache struct {
+	overLimit map[string]bool
+
+	// limit is the number of maximum number of IP per host.
+	limit int
```
`perHostLimit` will be more clear when using it later.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +181,124 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+
+// cleanupOverLimitEntries returns the names that has reached the max number of
+// IP per host.
+// Internally the function sort the entries by the TTL.
```
We should say sorted by expiration time. TTL might be ambiguous and mean the actual TTL seconds value.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +181,124 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
-func (c *DNSCache) UpdateFromCache(update *DNSCache) {
+// individually. If any item in the names slice will only update that names.
+// If the clean argument is set to true, all entries will be clean.
+func (c *DNSCache) UpdateFromCache(update *DNSCache, names []string, clean bool) {
```
names might be better as namesToUpdate or something,

the comment is a little unclear, maybe:
When names is non-nil only those names are updated from update, otherwise all DNS names in update are used.
When clean is true the data in c for each specific name is cleared before the update.

raybejjani commented on pkg/option/config.go:
```diff
@@ -260,6 +260,10 @@ const (
 
+	// ToFQDNsMaxIPsPerHost defines the maximum number of IPs to maintain
+	// for each FQDN name in an endpoint's FQDN cache
+	ToFQDNsMaxIPsPerHost = "tofqdns-max-ip-per-hostname"
```
`tofqdns-endpoint-max-ip-per-hostname` might be a little more clear but it is longer.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 // applies, thus the merged outcome is consistent with adding the entries
 // individually.
-func (c *DNSCache) UpdateFromCache(update *DNSCache) {
+// When namesToUpdate is non-nil only those names are updated from update, otherwise
```
I guess this isn't wholly true. "When namesToUpdate has non-zero length" might be more correct (since someone might pass in an empty list thinking it will be a no-op).

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -173,8 +306,19 @@ func (c *DNSCache) UpdateFromCache(update *DNSCache) {
 	defer c.Unlock()
-
-	for _, newEntries := range update.forward {
+	if len(namesToUpdate) == 0 {
```
I wonder if it's better to convert the string slice to a map when `len(namesToUpdate) != 0` instead. This way, the default case just uses update.forward without doing extra work. It probably won't matter too much in the end.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -173,8 +306,19 @@ func (c *DNSCache) UpdateFromCache(update *DNSCache) {
+			continue
+		}
+		if clean {
+			c.removeExpired(newEntries, time.Now(), time.Now())
```
I realised that this is wrong. c.removeExpired should only work on entries in c, and not from update. The effect here would be that removeReverse inside removeExpired etc. wouldn't work correctly.
There's also another problem with the assumption I made yesterday. If the first endpoint.DNSHistory doesn't have `name` then we won't ever clean it. Another problem is that the time.Now() here leaves a bit of a race.
So, I think we are better off not having clean in this function, but call ForceExpire before we call this function in the controller. This should keep things relatively consistent (at the cost of a regex compile).

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -173,8 +306,19 @@ func (c *DNSCache) UpdateFromCache(update *DNSCache) {
+			c.removeExpired(newEntries, time.Now(), time.Now())
+		}
 		for _, newEntry := range newEntries {
 			c.updateWithEntry(newEntry)
```
Not related with this PR, but could you add a comment like:
using newEntry from update violates encapsulation but these objects are immutable and so it should be harmless to do this (but still a bad idea).

Alternatively, we can copy newEntry (The type is shallow-copyable in general, unless we also play with IP slices, and if we do we deserve bugs).

raybejjani commented on pkg/fqdn/helpers.go:
```diff
@@ -165,3 +165,17 @@ func sortedIPsAreEqual(a, b []net.IP) bool {
 	return strings.ToLower(dns.Fqdn(matchName))
 }
+
+func KeepUniqueNames(names []string) []string {
```
I have a slight preference for only using the slice instead of making a map. In any case, however we do it we should document the function and explain that the output slice is the same or different.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+		ip    string
+		entry *cacheEntry
+	}
+	toClean := []string{}
```
This variable name is confusing because the function is called cleanupOverLimitEntries. affectedNames or something in the past tense is better.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+		}
+
+		sort.Slice(sortedEntries, func(i, j int) bool {
+			return sortedEntries[i].entry.ExpirationTime.Unix() < sortedEntries[j].entry.ExpirationTime.Unix()
```
This isn't wrong but `sortedEntries[i].entry.ExpirationTime.Before(sortedEntries[j].entry.ExpirationTime)` is a little more precise. It won't really ever be an issue, though.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -108,15 +108,39 @@ type DNSCache struct {
 
+// NewDNSCache returns an initialized DNSCache and set the max host limit to
+// the given argument
+func NewDNSCacheWithLimit(limit int) *DNSCache {
```
Is there a reason to make a second constructor? perhaps we can have a `NoOverlimit` const that would make it clear in `NewDNSCache(NoOverlimit)`

eloycoto replied:
for me is easy with two constructors, and it is been used in different ways.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 		c.forward[entry.Name] = entries
 	}
 	c.updateWithEntryIPs(entries, entry)
+	if len(entries) > c.perHostLimit {
```
Let's have this say `if c.perHostLimit > 0 && len(entries) > c.perHostLimit {` so overLimit is never filled in to begin with when it is disabled.

raybejjani replied:
Oh, also, this should go after c.removeExpired below, since after expiration we may not be over the limit.
Similarly, we should check if we are under the limit, and remove ourselves from overlimit:
```
switch {
  case c.perHostLimit > 0 && len(entries) > c.perHostLimit:
    c.overLimit[entry.Name] = true
  case c.perHostLimit > 0 && len(entries) < c.perHostLimit:
     delete(c.overLimit, entry.Name)
}

eloycoto replied:
Good one, the delete does not matter here. We already checked on the GC function ;-)

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
+	}
+	toClean := []string{}
+
+	// For global cache the limit maybe is not used at all.
```
Without knowing about the various instances of this type that we use, this is a little confusing. If you wanted to make sure it was clear, maybe add this comment where we instantiate `fqdn.DefaultDNSCache`?

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -266,8 +410,10 @@ func (c *DNSCache) updateWithEntryIPs(entries ipEntries, entry *cacheEntry) {
+
 }
 
 // removeExpired removes expired (or nil) cacheEntry pointers from entries, an
```
Since we know the entry we are expiring, we can also remove it from cleanup. This way if someone forgets to run GC we won't leak memory.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -108,15 +108,39 @@ type DNSCache struct {
+
+	// cleanup is the map of TTL expiration time and the entries that are going
+	// to expire will be deleted by CleanupExpiredEntries.
+	cleanup map[int64][]string
```
We should explain how cleanup and overlimit are being used more completely here. The difference between them is a little unclear. It's possible something like `expirationTimes` is a better name but it is longer.
For cleanup:
cleanup maps the TTL expiration times (in seconds since the epoch) to DNS names that expire in that second. On every new insertion where the new data is actually inserted into the cache (i.e. it expires later than an existing entry) cleanup will be updated. CleanupExpiredEntries cleans up these entries on demand.
Note: Lookup functions will not return expired entries, and this is used to proactively enforce expirations.
Note: It is important to periodically call CleanupExpiredEntries otherwise this map will grow forever.
(this last note isn't needed if we remove entries from cleanup during `removeExpired`)

For overlimit:
overlimit is a set of DNS names that were over the per-host configured limit when they received an update. The excess IPs will be removed when `cleanupOverLimitEntries` is called, but will continue to be returned by Lookup until then.
Note: It is important to periodically call GC otherwise this map will grow forever (it is very bounded, however).
(this last note isn't needed if we remove entries from overLimit when we insert IPs).

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +178,125 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 
+// AddNameToCleanup adds the IP with the given TTL to the the cleanup map to
+// delete the entry from the policy when it expires.
+// Need to be called with the lock
```
with a write lock (I think we make the distinction elsewhere)

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -108,15 +108,39 @@ type DNSCache struct {
+		reverse:     make(map[string]nameEntries),
+		lastCleanup: time.Now(),
+		cleanup:     map[int64][]string{},
+		overLimit:   map[string]bool{},
```
since we set every other field, perhaps we should also set perHostLimit. It is redundant but makes it clear that it was intended. shrug.

raybejjani commented on pkg/fqdn/cache.go:
```diff
@@ -154,16 +190,124 @@ func (c *DNSCache) updateWithEntry(entry *cacheEntry) {
 		c.forward[entry.Name] = entries
 	}
 	c.updateWithEntryIPs(entries, entry)
+	if c.perHostLimit > 0 && len(entries) > c.perHostLimit {
```
can you move this to after the removeExpired below?

eloycoto replied:
done!

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	entries, _ := cache.cleanupExpiredEntries(time.Now().Add(5 * time.Second))
+	c.Assert(len(entries), Equals, 1)
+	c.Assert(len(cache.cleanup), Equals, 0)
+	c.Assert(len(cache.forward["test.com"]), Equals, 0)
```
Could you switch this to use `.Lookup`? That would test the complete lookup behaviour. Also, could you also test `.LookupIP` to make sure the reverse lookup is correctly cleaned.

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	}
+	c.Assert(cache.cleanupOverLimitEntries(), checker.DeepEquals, []string{"test.com"})
+	c.Assert(len(cache.forward["test.com"]), Equals, limit)
+	c.Assert(cache.forward["test.com"]["1.1.1.1"], IsNil)
```
We should also test 1.1.1.2, because we consciously decided to deleted the older by expiration time. We should test that .Lookup returns the remaining 5 IPs, and that `.LookupIP` doesn't work for 1.1.1.{1,2} but does for the rest.

eloycoto replied:
There is no way that 1.1.1.2 need to be deleted no? 

```
	for i := 1; i < limit+2; i++ {
		cache.Update(now, "test.com", []net.IP{net.ParseIP(fmt.Sprintf("1.1.1.%d", i))}, i)
	}
```

raybejjani replied:
oh! you are correct. I read it as `<= limit+2` so yes, in the current case, you are correct and only 1.1.1.1 would be missing.

I wonder, though, if we should test that we remove more than one IP when it is over the limit. And what do we do if a DNS lookup has 4 IPs but we remove only 2?

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+		cache.Update(now, "test.com", []net.IP{net.ParseIP(fmt.Sprintf("1.1.1.%d", i))}, i)
+	}
+	c.Assert(cache.cleanupOverLimitEntries(), checker.DeepEquals, []string{"test.com"})
+	c.Assert(len(cache.forward["test.com"]), Equals, limit)
```
Could you switch all the .forward checks to use `.Lookup`?

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	c.Assert(cache.forward["test.com"]["1.1.1.1"], IsNil)
+	c.Assert(len(cache.forward["foo.bar"]), Equals, 1)
+	c.Assert(len(cache.forward["bar.foo"]), Equals, 1)
+	c.Assert(len(cache.overLimit), Equals, 0)
```
This might be better tested with another `c.Assert(len(cache.cleanupOverLimitEntries()), Equals, 0)` so we test that the next time we run `cleanupOverLimitEntries` it doesn't do something unexpected.

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
 	}
 }
+
+func (ds *DNSCacheTestSuite) TestCleanupEntries(c *C) {
```
We use TTLCleanup in the other test. We should use the same word to make it clear that they are the same thing (cleanup doesn't say anything about TTL so it is already unclear)

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+func (ds *DNSCacheTestSuite) TestGCOverlimitAfterTTLCleanup(c *C) {
+	limit := 5
+	cache := NewDNSCacheWithLimit(limit)
+	cache.lastCleanup = time.Now().Add(-1 * time.Minute)
```
We should explain why we're doing this, since it won't be clear later. Why is it needed, though? Update operates on the expiration time, and cleanupExpiredEntries takes a parameter directly.

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	for i := 1; i < limit+2; i++ {
+		cache.Update(now, "test.com", []net.IP{net.ParseIP(fmt.Sprintf("1.1.1.%d", i))}, 1)
+	}
+	c.Assert(len(cache.forward["test.com"]), Equals, limit+1)
```
We should use `.Lookup`, and `.LookupIP` in the assert.

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	c.Assert(cache.cleanupOverLimitEntries(), checker.DeepEquals, []string{})
+}
+
+func (ds *DNSCacheTestSuite) TestOverlimitAfterDeleteForwardEntry(c *C) {
```
This is a dangerous test. We test the implementation details of how we access `.forward` but what we really need to test is that after an entry is removed from from `.forward` `cleanupOverLimitEntries` doesn't crash, but also doesn't break anything. `TestGCOverlimitAfterTTLCleanup` does the second part. It might also do the first?

raybejjani commented on pkg/fqdn/cache_test.go:
```diff
@@ -464,3 +465,65 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
+	c.Assert(len(cache.forward["test.com"]), Equals, 4)
+}
+
+func (ds *DNSCacheTestSuite) TestGCOverlimitAfterTTLCleanup(c *C) {
```
It's a little wierd that we're not testing the actual GC function itself, but re-implement it sort-of.

eloycoto replied:
because of the time, if not I need to do a time.Sleep in the code.

raybejjani replied:
oh, but you have `cache.lastCleanup = time.Now().Add(-1 * time.Minute)` so you already set it up so that "now" will work correctly, no?

raybejjani reviewed (approved):
Some comments on the tests, nothing blocking. Where I say to use `.Lookup` and `.LookupIP` I really mean the `*WithTime` variant used in the other tests.

eloycoto committed:
FQDN Rulegen: Fix issue if ips already expired

On `updateIPsForName` if the new data has already expired or the old
data also expired can be issue in regeneration.

Fix #6796

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon reviewed (approved):
LGTM for CLI
